### DISCUSSION TITLE
About cache write when using ISR

### DISCUSSION THREAD

**Discussion by ya2s** [vercel/next.js#69177] [Help] [answered]
### Summary

I'm using ISR with pages router.
When validating data on origin after the time set in revalidate, if the data is the same, won't cache write occur?
In other words, if the props of getStaticProps are always the same, won't cache write occur?

**Comment by suprunchuk** [comment, marked as answer]
When using ISR with getStaticProps and the revalidate option, here's what happens:

1. The page is initially built and cached at build time.
2. When a request comes in after the revalidate time has passed, Next.js will still serve the cached page.
3. In the background, Next.js triggers a regeneration of the page.
4. During this regeneration, getStaticProps is called again to fetch new data.
5. If the new data is different from the cached data, the page is regenerated and the cache is updated.
6. If the new data is the same as the cached data, the behavior depends on the Next.js version:

    - In Next.js versions prior to 12.2.0, a cache write would occur even if the data hasn't changed.
    - In Next.js 12.2.0 and later, an optimization was introduced to avoid unnecessary cache writes when the data hasn't changed.

For Next.js 12.2.0 and later:

If the props returned by getStaticProps are exactly the same as the cached version (determined by a deep comparison), Next.js will not trigger a cache write. This optimization helps reduce unnecessary builds and improves performance.

Here's an example to illustrate:

```js
export async function getStaticProps() {
  const data = await fetchSomeData();
  return {
    props: { data },
    revalidate: 60, // Revalidate every 60 seconds
  };
}
```

**Reply by ya2s** [reply]
Thank you for explaining in detail😆

> In Next.js 12.2.0 and later, an optimization was introduced to avoid unnecessary cache writes when the data hasn't changed.
For Next.js 12.2.0 and later:
If the props returned by getStaticProps are exactly the same as the cached version (determined by a deep comparison), Next.js will not trigger a cache write. This optimization helps reduce unnecessary builds and improves performance.

Is this behavior described somewhere in the documentation?

**Reply by suprunchuk** [reply]
This particular optimization was introduced as part of the improvements in Next.js 12.2.0. While it's not prominently featured in the main documentation, you can find references to it in:

- https://vercel.com/blog/vercel-cache-api-nextjs-cache
- https://www.smashingmagazine.com/2021/04/incremental-static-regeneration-nextjs/
